File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
      public int maxSubArray(int[] nums) {
        int n = nums.length;
        
        // Step 1: Create DP array
        int[] dp = new int[n];
        
        // Step 2: Base case
        dp[0] = nums[0];
        
        // Step 3: Fill DP table
        for (int i = 1; i < n; i++) {
            dp[i] = Math.max(nums[i], nums[i] + dp[i - 1]);
        }
        
        // Step 4: Find global maximum
        int maxSum = dp[0];
        for (int i = 1; i < n; i++) {
            maxSum = Math.max(maxSum, dp[i]);
        }
        
        return maxSum;
      }
}

File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_value = nums[0]
        cur_sum = 0

        for i in nums:
            if cur_sum < 0:
                cur_sum = 0
            cur_sum += i
            
            max_value = max(max_value, cur_sum)
        return max_value
            

        

            

                
        